### PR TITLE
improve test coverage

### PR DESCRIPTION
**Pull request by wxtim**
Partial Response to https://github.com/cylc/cylc-rose/issues/157, and partial replacement of https://github.com/cylc/cylc-rose/pull/196

The aim from my point of view is to separate easy small changes from the need to refactor subprocess tests more thoroughly to work in a manner similar to Cylc's integration tests.

My checks on coverage suggest that this brings coverage from just over 95% to just under 99%

**Check List**

- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to both `setup.cfg` (and `conda-environment.yml` if present).
- [x] This a change to tests which should not require documentation.

### CODE REVIEW

File: cylc/rose/entry_points.py [hunk 1]
```diff
@@ -71,8 +71,7 @@ def post_install(srcdir=None, opts=None, rundir=None):
         srcdir=srcdir, opts=opts, rundir=rundir
     )
     # Finally dump a log of the rose-conf in its final state.
-    if results['fileinstall']:
-        dump_rose_log(rundir=rundir, node=results['fileinstall'])
+    dump_rose_log(rundir=rundir, node=results['fileinstall'])
```
Comment (MetRonnie) [inline cylc/rose/entry_points.py:74]: Without type checking I'm not confident in this change. It looks to me like the `node` argument shouldn't be a `False` or `None` so I think the truthiness test was valid?

File: cylc/rose/entry_points.py [hunk 2]
```diff
@@ -177,11 +176,6 @@ def record_cylc_install_options(
     """
     # Create a config based on command line options:
     cli_config = get_cli_opts_node(opts, srcdir)
-    # Leave now if there is nothing to do:
-    if not cli_config:
-        return False
```
Comment (MetRonnie) [inline cylc/rose/entry_points.py:182]: I think this ok, looks like `get_cli_opts_node()` always returns an object

File: cylc/rose/entry_points.py [hunk 3]
```diff
@@ -177,11 +176,6 @@ def record_cylc_install_options(
...
-    # raise error if CLI config has multiple templating sections
-    identify_templating_section(cli_config)
```
Comment (MetRonnie) [inline cylc/rose/entry_points.py:184]: Why has this been removed?

File: cylc/rose/entry_points.py [hunk 4]
```diff
@@ -218,8 +212,7 @@ def record_cylc_install_options(
     dumper.dump(cli_config, str(conf_filepath))
 
     # Merge the opts section of the rose-suite.conf with those set by CLI:
-    if not rose_conf_filepath.is_file():
-        rose_conf_filepath.touch()
+    rose_conf_filepath.touch()
```
Comment (MetRonnie) [inline cylc/rose/entry_points.py:215]: Not sure about this one. The next line loads this file so if it is always a newly created empty file what would be the point of loading it?

**Review by MetRonnie — COMMENTED** [review]
Partial review

File: cylc/rose/entry_points.py (hunk 1 above)
Comment (wxtim) [reply cylc/rose/entry_points.py:74]: https://github.com/cylc/cylc-rose/pull/196/files#r1033649794

File: cylc/rose/entry_points.py (hunk 2 above)
Comment (wxtim) [reply cylc/rose/entry_points.py:182]: https://github.com/cylc/cylc-rose/pull/196/files#r1033650767

File: cylc/rose/entry_points.py (hunk 4 above)
Comment (wxtim) [reply cylc/rose/entry_points.py:215]: https://github.com/cylc/cylc-rose/pull/196/files#r1033651643

File: cylc/rose/entry_points.py (hunk 1 above)
Comment (MetRonnie) [reply cylc/rose/entry_points.py:74]: https://github.com/cylc/cylc-rose/blob/2a68e4c1438400e81fd44dad2a84bbff70d46630/cylc/rose/entry_points.py#L227-L237

Comment (wxtim) [reply cylc/rose/entry_points.py:74]: This has been hanging around too long. A quick re-read suggests you are right.

```suggestion
    if results['fileinstall']:
        dump_rose_log(rundir=rundir, node=results['fileinstall'])
```

Comment (wxtim) [inline cylc/rose/entry_points.py:74]:
```suggestion
    if results['fileinstall']:
        dump_rose_log(rundir=rundir, node=results['fileinstall'])
```

File: tests/unit/test_rose_opts.py [hunk 5]
```diff
@@ -86,8 +86,7 @@ def test_rose_fileinstall_validate(fixture_provide_flow, cylc_validate_cli):
 def test_rose_fileinstall_run(fixture_install_flow):
     """Workflow installs:
     """
-    _, _, _, result, _ = fixture_install_flow
-    assert result.ret == 0
+    fixture_install_flow
```
Comment (oliver-sanders) [inline tests/unit/test_rose_opts.py:89]:
```suggestion
    pass  # this tests the fixture itself
```

**Review by oliver-sanders — APPROVED** [review]

File: cylc/rose/utilities.py [hunk 6]
```diff
@@ -454,9 +444,7 @@ def get_cli_opts_node(opts=None, srcdir=None):
         )
 
     # Specialised treatement of optional configs.
-    if 'opts' not in newconfig:
-        newconfig['opts'] = ConfigNode()
-        newconfig['opts'].value = ''
+    newconfig['opts'].value = ''
```
Comment (MetRonnie) [inline cylc/rose/utilities.py:447]: Unnecessary?
```suggestion
```

File: tests/unit/test_rose_stem_units.py [hunk 7]
```diff
@@ -297,6 +309,56 @@ def test__deduce_mirror():
...
+    stemrunner = get_StemRunner({}, options={'defines': []})
+    monkeypatch.setattr(
+        cylc.rose.stem.StemRunner, '_read_auto_opts',
+        lambda _: 'foo=bar baz=qux=quiz'
+    )
+    stemrunner._parse_auto_opts()
+    assert 'foo="bar"' in stemrunner.opts.defines[0]
+
+
+def test_process_no_auto_opts(monkeypatch, get_StemRunner):
```
Comment (MetRonnie) [inline tests/unit/test_rose_stem_units.py:352]: Docstring?

File: cylc/rose/utilities.py (hunk 6 above)
Comment (wxtim) [reply cylc/rose/utilities.py:447]: No. Try commenting it out and see what happens to the tests: The  `merge_opts` function requires the `newconfig['opts']` object to have something assigned to the value attribute.

File: tests/unit/test_rose_stem_units.py [hunk 8]
```diff
@@ -297,6 +309,56 @@ def test__deduce_mirror():
...
+            'url: https://foo.com/helloworld\n'
+            'project: helloworld\n'
+        ),
+        'stderr'
+    ))})
+    result = stemrunner._ascertain_project('')
+    assert result == ('foo', '', '', '', 'foo')
+
+
+def test_process_multiple_auto_opts(monkeypatch, get_StemRunner):
```
[the remark this answers is not in the document]
Comment (wxtim) [reply tests/unit/test_rose_stem_units.py]: I just test the code as is, it took me a few minutes to work out what it was trying to do. Now I don't like what it's trying to do: It doesn't raise an error if there is a mistake in the Rose user/global config.

**Review by MetRonnie — APPROVED** [review]

**Pull request merged**